cpp/src/arrow/util/functional.h

```diff
@@ -0,0 +1,59 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <tuple>
+
+#include "arrow/util/macros.h"
+
+namespace arrow {
+
+/// Helper struct for examining lambdas and other callables.
+/// If the callable is not overloaded, the argument types of its call operator can be
+/// extracted via single_call::argument_type<Index, Function>
+struct single_call {
```

Comment:
Why is it called `single_call`? Can I only use it a single time? Perhaps call it `call_traits`?

Reply:
Perhaps `call_traits<Function>::argument_type<Index>` would be a nicer spelling btw. No strong feelings though.

Reply:
I like this `call_traits<Function>::argument<Index>` and `call_traits<Function>::argument<Index>`.

Reply:
It's called `single_call` because it only works for function objects which are not overloaded. `non_overloaded_call_traits`?

The alternate spelling is doable but would be `call_traits<Function>::template argument<Index>`, which I don't think is prettier.

Reply:
Not very pretty indeed. `call_traits` sounds fine to me. I don't know if there's a way of outputting a clear error message if it's called on an overloaded function?

Reply:
Ideally, one uses `call_traits::check` (returns true for non-overloaded) or `call_traits::enable_if` (sfinae if not `check`) to restrict usage of the other members to non overloaded functions